hazelcast/src/main/java/com/hazelcast/internal/diagnostics/EventQueuePlugin.java

```diff
@@ -0,0 +1,242 @@
+        this.threshold = props.getInteger(THRESHOLD);
+        this.samples = props.getInteger(SAMPLES);
+    }
+
+    @Override
+    public long getPeriodMillis() {
+        return periodMillis;
+    }
+
+    @Override
+    public void onStart() {
+        logger.info("Plugin:active, period-millis:" + periodMillis + " threshold:" + threshold + " samples:" + samples);
+    }
+
+    @Override
+    public void run(DiagnosticsLogWriter writer) {
+        writer.startSection("EventQueues");
+
+        int index = 1;
+        List<BlockingQueue<Runnable>> eventQueues = getEventQueues();
+        for (BlockingQueue<Runnable> eventQueue : eventQueues) {
+            clear();
+            scan(writer, eventQueue, index++);
+        }
+
+        writer.endSection();
+    }
+
+    // just for testing
+    ItemCounter<String> getOccurrenceMap() {
+        return occurrenceMap;
+    }
+
+    private List<BlockingQueue<Runnable>> getEventQueues() {
+        return eventExecutor.getWorkQueues();
+    }
+
+    private void scan(DiagnosticsLogWriter writer, BlockingQueue<Runnable> eventQueue, int index) {
+        int sampleCount = sample(eventQueue);
+        if (sampleCount < 0) {
+            return;
+        }
+
+        render(writer, sampleCount, index);
+    }
+
+    private void render(DiagnosticsLogWriter writer, int sampleCount, int index) {
+        writer.startSection("worker=" + index);
+
+        writer.writeKeyValueEntry("eventCount", events.size());
+        writer.writeKeyValueEntry("sampleCount", sampleCount);
+        renderSamples(writer, sampleCount);
+
+        writer.endSection();
+    }
+
+    private void renderSamples(DiagnosticsLogWriter writer, int sampleCount) {
+        writer.startSection("samples");
+
+        for (String key : occurrenceMap.keySet()) {
+            long value = occurrenceMap.get(key);
+            if (value == 0) {
+                continue;
+            }
+
+            double percentage = (1d * value) / sampleCount;
+            writer.writeEntry(key + " sampleCount=" + value + " " + defaultFormat.format(percentage));
+        }
+        writer.endSection();
+    }
+
+    private void clear() {
+        occurrenceMap.reset();
+        events.clear();
+    }
+
+    /**
+     * Samples the queue.
+     *
+     * @param queue the queue to sample
+     * @return the number of samples, or -1 if there were not sufficient samples
+     */
+    private int sample(BlockingQueue<Runnable> queue) {
+        for (Runnable runnable : queue) {
+            events.add(runnable);
+        }
+
+        int eventCount = events.size();
+        if (eventCount < threshold) {
+            return -1;
+        }
+
+        int sampleCount = min(samples, eventCount);
+        int actualSampleCount = 0;
+        while (actualSampleCount < sampleCount) {
+            Runnable runnable = events.get(random.nextInt(eventCount));
+            actualSampleCount += sampleRunnable(runnable);
+        }
+
+        return actualSampleCount;
```

Comment:
Hm, should we `clear()` the events here at the end to avoid keeping hard references from the plugin? Do we even need to have `events` as a field or just as a parameter and variable to render the samples?